thetis/inversion_tools.py

```diff
@@ -12,37 +12,64 @@
 import os
 
 
-class OptimisationProgress(object):
+class InversionManager(FrozenHasTraits):
     """
-    Class for stashing progress of the optimisation routine.
+    Class for handling inversion problems and stashing
+    the progress of the associated optimization routines.
     """
-    J = 0  # cost function value (float)
-    dJdm_list = None  # cost function gradient (Function)
-    m_list = None  # control (Function)
-    m_progress = []
-    J_progress = []
-    dJdm_progress = []
-    i = 0
-    tic = None
-    nb_grad_evals = 0
-    control_coeff_list = []
-    control_list = []
-
-    def __init__(self, output_dir='outputs', no_exports=False, real=False):
+
+    @unfrozen
+    def __init__(self, sta_manager, output_dir='outputs', no_exports=False, real=False,
+                 penalty_parameters=[], cost_function_scaling=None,
+                 test_consistency=True, test_gradient=True):
         """
+        :arg sta_manager: the :class:`StationManager` instance
         :kwarg output_dir: model output directory
         :kwarg no_exports: toggle exports to vtu
         :kwarg real: is the inversion in the Real space?
-        """
+        :kwarg penalty_parameters: a list of penalty parameters to pass
+            to the :class:`ControlRegularizationManager`
+        :kwarg cost_function_scaling: global scaling for the cost function.
+            As rule of thumb, it's good to scale the functional to J < 1.
+        :kwarg test_consistency: toggle testing the correctness with
+            which the :class:`ReducedFunctional` can recompute values
+        :kwarg test_gradient: toggle testing the correctness with
+            which the :class:`ReducedFunctional` can recompute gradients
+        """
+        assert isinstance(sta_manager, StationObservationManager)
+        self.sta_manager = sta_manager
+        self.reg_manager = None
         self.output_dir = output_dir
         self.no_exports = no_exports or real
         self.real = real
+        self.penalty_parameters = penalty_parameters
+        self.cost_function_scaling = cost_function_scaling
+        self.sta_manager.cost_function_scaling.assign(cost_function_scaling)
```

Comment:
`cost_function_scaling` is `None` by default which will raise an error in the assign. If `None` use `Constant(1)`?